### DISCUSSION TITLE
offline install nix neovim saved in other machine via usbstick？

### DISCUSSION THREAD

**Discussion by qlibp** [nix-community/kickstart-nix.nvim#18] [Q&A] [answered]
How could I save neovim stuff within /nix/store into a tarball file and then load it in other machines offline？

**Comment by qlibp** [comment]
already try nix copy cmd, but it says source tree can’t find /flake.nix. But I find that flake file in /nix/store

**Reply by mrcjkb** [reply]
:thinking: I just tried `nix copy .#nvim --no-check-sigs --to /tmp/kickstart-nix` and it worked. What exactly did you run?

**Reply by qlibp** [reply]
1. I modify some nix & nvim config, and then run `nix build`, after that I perform `nix copy .#nvim --to /tmp/cache`
2. I mount the /tmp/cache into a clean docker container to perform offline installation test.
3. I run `nix copy --from /tmp/cache .#nvim`, and got  source tree can’t find /flake.nix in the docker container.

note: I may post a screenshot later. Thx for your reply!

**Reply by mrcjkb** [reply]
nix copy copies the store path of a derivation, not its source. That's why you're seeing the error on `nix copy --from /tmp/cache .#nvim`

Instead of `.#nvim`, you need to pass the nix store URI: `--from file:///tmp/cache/nix/store/...`

**Reply by qlibp** [reply]
not sure why I got no /tmp/cache/nix/store in my path. 

Here is the screenshoot I got. I don’t know what nix copy is actually doing, does it dump all binary? or is it just dump the metadata? I still need to download the binary from the internet?
![IMG_844226AAC4E7-1](https://github.com/nix-community/kickstart-nix.nvim/assets/43082379/6ad4dd1d-c36e-46cd-bc05-93f45a845fc2)

**Reply by mrcjkb** [reply]
What do you end up with in `/tmp/cache` when you run `nix copy .#nvim --to /tmp/cache` (or `nix copy .#nvim --to file:///tmp/cache`)?

According to [the docs](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-copy), the behaviour should depend on whether or not you have a `file://` prefix.

**Reply by qlibp** [reply]
I use `nix copy .#nvim --to file:///tmp/cache`, `nix copy .#nvim --to /tmp/cache` will fail

**Reply by qlibp** [reply]
Aditional info, `nix copy .#nvim --from file:///tmp/cache --debug` 
<img width="1371" alt="image" src="https://github.com/nix-community/kickstart-nix.nvim/assets/43082379/3435c8fc-d4c1-4352-9148-7a67bcd40812">

Does it mean we can only use `nix copy .#nvim --to /tmp/cache`?

**Reply by mrcjkb** [reply]
No. Afaik, you can't pass `.#nvim` when copying **from** a binary cache. You need the full nix store path (which is the same path you get when you run `nix eval --raw .#nvim`), or after a `nix build`.

So it should be something like 

```
nix copy --from file:///tmp/cache /nix/store/<hash>-nvim...`

**Reply by qlibp** [reply]
not sure why, you could see my cmd in the screenshot, not sure why getting such signature error
<img width="1387" alt="image" src="https://github.com/nix-community/kickstart-nix.nvim/assets/43082379/6a3f8f54-dc2e-4e5b-a8eb-1d97b4251b17">

**Reply by mrcjkb** [reply]
https://github.com/NixOS/nix/issues/4894#issuecomment-855401632

**Reply by qlibp** [reply]
already use no-check-sigs when calling nix copy to, should I use this flag when I call nix copy from?

**Reply by mrcjkb** [reply]
Yes, probably. 
I'm not an expert on nix copy, so you'll have to do some digging yourself.

**Reply by qlibp** [reply]
yep, both nix copy from and nix copy to needs --no-check-sigs, all stuff get copied inside /nix/store now. But when I try to `nix run` in kickstarter's repo, it tries to download stuff from the internet, how can I tell flake to use the local /nix/store cache without downloading anything else?

**Reply by mrcjkb** [reply]
https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run#miscellaneous-global-options

**Reply by qlibp** [reply]
have tried --offline flag, it will still download nixpkg and flake stuff from internet. Is nixpkg and flake stuff not copied in using nix copy? Not sure what's get copied using nix copy. Thx for your help!

```
root@b7c323601b36:~/test_nix_with_nvim/kickstart-nix.nvim# nix run --offline .#nvim
error:
       … while calling the 'import' builtin
         at /nix/store/c717ywya6y80b89rn9gwr7haa7lhqzym-source/flake.nix:35:14:
           34|     flake-utils.lib.eachSystem supportedSystems (system: let
           35|       pkgs = import nixpkgs {
             |              ^
           36|         inherit system;

       … while realising the context of a path

       … in the condition of the assert statement
         at «nix-internal»/call-flake.nix:66:13:
           65|           if node.flake or true then
           66|             assert builtins.isFunction flake.outputs;
             |             ^
           67|             result

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: unable to download 'https://github.com/NixOS/nixpkgs/archive/44d0940ea560dee511026a53f0e2e2cde489b4d4.tar.gz': Timeout was reached (28)

```

**Reply by mrcjkb** [reply, marked as answer]
`nix copy --from` should copy the derivation to the nix store path, so you should be able to run `/nix/store/.../bin/nvim`. There's no need for `nix run`.

**Reply by qlibp** [reply]
oh wow! It works! I think I should start analysing how it actually works. It surprise me that /nix/store/xxx-neovim can load config from the right place and call pylsp within/nix/store. Though I know that the flake specify the rtp of neovim, but I still can’t figure out the whole picture.

Anyway, thanks a lot!

**Reply by mrcjkb** [reply]
That's because `nix copy` copies all the nix store paths that the `nvim` nix store path transitively depends on :smile: 
Glad to hear it's working for you!

**Reply by qlibp** [reply]
Hey there, would you mind elaborate more for me on the reason why nix run doesn’t work here?

If i just symlink the neovim binary, it seems that I have no way to use other nix cmd to activate a dev env?

**Reply by mrcjkb** [reply]
I don't understand the question.

**Reply by qlibp** [reply]
For now, if I use `nix copy` to copy the neovim related stuff to another machine, I can not use `nix run .#nvim` in pure offline mode to start the nvim app, do you know the reason why this happens?

It seems that some nixpkgs requires internet installation, even though I pass in  `--offline` flag. Isn't nix copy already copies the whole closure of neovim?

**Reply by mrcjkb** [reply]
`nix run` is for building and running a flake output. 
`nix copy` copies the nix store paths of the derivation, not the flake. You can't use `nix run` to run something you copied with `nix copy`. You need to use the binary in the nix store path that you copied.

**Reply by qlibp** [reply]
But what if I run `nix run`, will the flake reuse the copied derivation in `/nix/store`?

When I copy the clousure of the neovim derivation, what else is missing so that I need to download from the internet in order to make `nix run` work?

**Reply by mrcjkb** [reply]
> But what if I run nix run, will the flake reuse the copied derivation in /nix/store?

You mean within the image? 
I think you have to add the location to your binary caches. 

This is kind of a general question, not directly related to kickstart-nix.nvim. You're more likely to get help if you ask on the [NixOS discourse](https://discourse.nixos.org/c/learn).
